### PR TITLE
[JENKINS-43055] Add support for FlowExecutionListener

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-43055](https://issues.jenkins-ci.org/browse/JENKINS-43055)

Downstream of https://github.com/jenkinsci/workflow-api-plugin/pull/36

cc @reviewbybees

### CODE REVIEW

**Comment by abayer** [comment]
Still needs tests, but I wanted to see if the design was alright before proceeding.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 1]
```diff
@@ -489,6 +489,7 @@ public void start() throws IOException {
         g.register(s);
         final SettableFuture<CpsThreadGroup> f = SettableFuture.create();
         programPromise = f;
+
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: ?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java [hunk 2]
```diff
@@ -232,13 +233,20 @@ public void unexport(BodyReference ref) {
                 public Void call() throws Exception {
                     Jenkins j = Jenkins.getInstance();
                     if (paused.get() || j == null || j.isQuietingDown()) {
+                        for (FlowExecutionListener listener : FlowExecutionListener.all()) {
+                            listener.onPaused(execution);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java]: This is called whenever we are asked to schedule something, not necessarily when the pause button is pressed.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java (hunk 1 above)
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Dunno what happened there. Fixing.

**Comment by jglick** [comment]
Can you merge with `master` please so we can get a new build using yarn, which I hope will be more stable than npm?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java [hunk 3]
```diff
@@ -239,6 +241,10 @@ public Void call() throws Exception {
                         return null;
                     }
 
+                    for (FlowExecutionListener listener : ExtensionList.lookup(FlowExecutionListener.class)) {
+                        listener.onRunning(execution);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java]: This is going to be called many times during a build: every time we reënter the CPS VM thread. That does not match the expectations given by its Javadoc, which is one call at build start time and one call per running build per Jenkins restart.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java [hunk 4]
```diff
@@ -382,6 +384,53 @@ private static void assertStepExecutions(FlowExecution e, String... steps) throw
         }
     }
 
+    @Test
+    public void flowExecutionListener() throws Exception {
+        story.addStep(new Statement() {
+                          @Override
+                          public void evaluate() throws Throwable {
+                              WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                              p.setDefinition(new CpsFlowDefinition("echo 'Running for listener'; semaphore 'wait'", true));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java]: Try inserting a `sleep 0` in the middle here and I am betting you will get a count > 2.

Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java]: Nope, still got 2...

Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java]: Ok, threw in another sleep after the semaphore and did get 3.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java (hunk 3 above)
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java]: So where would be the right place to put this?

Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java]: Ah, think I figured it out.

**Comment by jglick** [comment]
> So where would be the right place to put this?

`CpsFlowExecution.start` for a new flow vs. `onLoad` for a resumed one (which I guess should just be distinct listener methods).

It is not actually obvious to me that the listener calls should be coming from `workflow-cps` at all. It is `WorkflowRun` which calls `FlowExecution.start` and `.onLoad`, so it would seem logical for `WorkflowRun` to be notifying listeners that these methods have been called. Similarly, `WorkflowRun.finish` would be a logical place to notify listeners that a flow has completed.

**Comment by abayer** [comment]
Yeah, that makes sense.

**Comment by jglick** [comment]
To be closed?

**Comment by abayer** [comment]
Yup, closing.

**Pull request closed without merging**